experimental/examples/imageviewer/shared/src/commonMain/kotlin/example/imageviewer/model/ScalableState.kt

```diff
@@ -4,81 +4,88 @@ import androidx.compose.runtime.getValue
+    ) {
+        this.areaSize = areaSize
+        this.targetSize = targetSize
+        applyLimits()
     }
 
-fun ScalableState.changeBoxSize(size: IntSize) {
-    boxSize = size
-    updateOffsetLimits()
-}
-
-fun ScalableState.setScale(scale: Float) {
-    this.scale = scale
-}
-
-fun ScalableState.addScale(diff: Float) {
-    scale = if (scale + diff > MAX_SCALE) {
-        MAX_SCALE
-    } else if (scale + diff < MIN_SCALE) {
-        MIN_SCALE
-    } else {
-        scale + diff
+    private fun applyLimits() {
+        if (targetSize.isSpecified && areaSize.isSpecified) {
+            offsetXLimits = centerLimits(targetSize.width * scale, areaSize.width)
+            offsetYLimits = centerLimits(targetSize.height * scale, areaSize.height)
+            offset = Offset(
+                offset.x.coerceIn(offsetXLimits),
+                offset.y.coerceIn(offsetYLimits),
+            )
+        } else {
+            offsetXLimits = Float.NEGATIVE_INFINITY..Float.POSITIVE_INFINITY
+            offsetYLimits = Float.NEGATIVE_INFINITY..Float.POSITIVE_INFINITY
+        }
     }
-    updateOffsetLimits()
-}
-
-fun ScalableState.addDragAmount(diff: Offset) {
-    offset -= IntOffset((diff.x + 1).toInt(), (diff.y + 1).toInt())
-    updateOffsetLimits()
-}
-
-fun ScalableState.updateImageSize(width: Int, height: Int) {
-    imageSize = IntSize(width, height)
-    updateOffsetLimits()
-}
 
-private fun ScalableState.updateOffsetLimits() {
-    if (offset.x + visiblePart.width > imageSize.width) {
-        changeOffset(x = imageSize.width - visiblePart.width)
-    }
-    if (offset.y + visiblePart.height > imageSize.height) {
-        changeOffset(y = imageSize.height - visiblePart.height)
+    private fun centerLimits(imageSize: Float, areaSize: Float): ClosedFloatingPointRange<Float> {
+        val areaCenter = areaSize / 2
+        val imageCenter = imageSize / 2
+        val extra = (imageCenter - areaCenter).coerceAtLeast(0f)
+        return -extra / 2..extra / 2
     }
-    if (offset.x < 0) {
-        changeOffset(x = 0)
+
+    fun addPan(pan: Offset) {
+        offset += pan
+        applyLimits()
     }
-    if (offset.y < 0) {
-        changeOffset(y = 0)
+
+    /**
+     * @param focus on which point the camera is focused in the area coordinate system.
+     * After we apply the new scale, the camera should be focused on the same point in
+     * the target coordinate system.
+     */
+    fun addScale(scaleMultiplier: Float, focus: Offset = Offset.Zero) {
+        setScale(scale * scaleMultiplier, focus)
     }
-}
 
-private fun ScalableState.changeOffset(x: Int = offset.x, y: Int = offset.y) {
-    offset = IntOffset(x, y)
+    fun setScale(scale: Float, focus: Offset = Offset.Zero) {
+        val newScale = scale.coerceIn(scaleLimits)
+        val focusInTargetSystem = (focus - offset) / this.scale
+        // calculate newOffset from this equation:
+        // focusInTargetSystem = (focus - newOffset) / newScale
```

Comment:
this commented line is needed for the future?

Reply:
It is not a commented line, but a part of a comment.

It is an equation that explains how we calculate newOffset in the next line.